{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ec1ca924-e742-4ddf-b865-f3f794dae15f",
   "metadata": {},
   "source": [
    "**16S V1-V9 diff kmers**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d4bd01fe-d6da-4b44-acec-8dcf8bac8ec9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import plotly.express as px\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from os import listdir\n",
    "from pandas import read_csv, DataFrame\n",
    "from tqdm import tqdm\n",
    "from subprocess import call\n",
    "from Bio.SeqIO import parse\n",
    "from skbio.stats.composition import clr \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec82cd14-3bc0-4ae0-8ce2-482378529645",
   "metadata": {},
   "outputs": [],
   "source": [
    "# one-time install\n",
    "pip install -U kaleido"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dec33e85-0385-4a5d-9697-36b095d432c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_fasta(output, mearged_pike_out, dbpath):\n",
    "    # Create fasta file \n",
    "    consensus = {}\n",
    "    cons_conter = 0\n",
    "    \n",
    "    with open(f'{output}/all_consensus.fasta', 'w') as opn_fasta:\n",
    "        for cons in mearged_pike_out.index:\n",
    "    \n",
    "            opn_fasta.write(f'>{cons_conter}\\n{cons}\\n')\n",
    "            consensus[cons_conter] = cons\n",
    "            cons_conter += 1\n",
    "    \n",
    "    return consensus\n",
    "    \n",
    "def run_blast(base, path):\n",
    "    \n",
    "    call(f'makeblastdb -in {base} -dbtype nucl', shell=True)\n",
    "    call(f'blastn -num_threads 60  -outfmt \"7 qseqid sseqid pident evalue qcovs bitscore\" -query {path}/all_consensus.fasta  -db {base} -out {path}/blast_results.txt', shell=True)\n",
    " #   pass\n",
    "def decode_tax(base) -> dict:\n",
    "    \n",
    "    # DB decoder \n",
    "    # Use db header format: Kingdom    Phylum    Class    Order    Family    Genus    Species\n",
    "    \n",
    "    base = parse(base, 'fasta')\n",
    "    taxonomy_linage = ['Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species']\n",
    "    tax_decoder = {}\n",
    "    \n",
    "    for line in tqdm(base):\n",
    "        \n",
    "        tax_decoder[line.id] = {}\n",
    "        linage = line.description.split(';')\n",
    "        linage[0] = linage[0].split()[1]\n",
    "    \n",
    "        for i in range(len(taxonomy_linage)):\n",
    "            try:\n",
    "                if taxonomy_linage[i] != 'Species':\n",
    "                \n",
    "                    tax_decoder[line.id][taxonomy_linage[i]] = linage[i]\n",
    "                \n",
    "                else:\n",
    "                    #print(linage[i].split())\n",
    "                    tax_decoder[line.id][taxonomy_linage[i]] = ' '.join(linage[i].split()[:2])\n",
    "                  \n",
    "            except:\n",
    "                \n",
    "                tax_decoder[line.id][taxonomy_linage[i]] = 'NA'\n",
    "    \n",
    "    return tax_decoder\n",
    "\n",
    "def parse_blast(path, \n",
    "                base, \n",
    "                data_tax, \n",
    "                consensus, \n",
    "                identity_filter, \n",
    "                cov_lim, \n",
    "                evalue_filter):\n",
    "    \n",
    "    # parser of blast table\n",
    "    \n",
    "    blast_header = ['qseqid',\n",
    "                    'sseqid', \n",
    "                    'pident',\n",
    "                    'evalue',\n",
    "                    'qcovs', \n",
    "                    'bitscore']\n",
    "    \n",
    "    blasting_results = {}\n",
    "    opn_blast = read_csv(f'{path}/blast_results.txt', sep='\\t', comment='#', header=None, names=blast_header)\n",
    "    \n",
    "    for i in tqdm(opn_blast['qseqid'].unique()):\n",
    "        \n",
    "        blast_subset = opn_blast[opn_blast[\"qseqid\"] == i]\n",
    "        blast_subset = blast_subset[blast_subset['pident'] >= identity_filter]\n",
    "        blast_subset = blast_subset[blast_subset['evalue'] <= evalue_filter]\n",
    "        blast_subset = blast_subset[blast_subset['qcovs'] >= cov_lim]\n",
    "\n",
    "        blast_subset = blast_subset.sort_values(by='evalue')\n",
    "        blast_subset = blast_subset.sort_values(by='pident')[::-1]\n",
    "\n",
    "        if len(blast_subset['sseqid'].values) == 0:\n",
    "            continue\n",
    "            \n",
    "        subject = blast_subset['sseqid'].values[0]\n",
    "        blasting_results[consensus[i]] = data_tax[subject]\n",
    "        \n",
    "    blasting_results_df = DataFrame(blasting_results).T\n",
    "    \n",
    "    return blasting_results_df\n",
    "    \n",
    "def processing_data_tax(data_tax):\n",
    "\n",
    "    data_tax_df = DataFrame(data_tax).T.fillna(0)\n",
    "    # Add pseudocunt\n",
    "    # data_tax_df = data_tax_df + 1\n",
    "    data_tax_df = data_tax_df.assign(m=data_tax_df.mean(axis=1)).sort_values('m').drop('m', axis=1)[np.sort(data_tax_df.columns)]\n",
    "\n",
    "    return  data_tax_df\n",
    "    \n",
    "def get_taxonomy(data_tax, \n",
    "                 blasting_results_df, \n",
    "                 mearged_pike_out, \n",
    "                 tax_level='OTU'):\n",
    "    \n",
    "    data_tax = {}\n",
    "    avs = np.intersect1d(blasting_results_df.index, mearged_pike_out.index)\n",
    "    count = 1\n",
    "    OTU_decoder  = {'Seq': [], 'OTU_name' : []}\n",
    "    \n",
    "    for av in tqdm(avs):\n",
    "\n",
    "        if tax_level == 'OTU':\n",
    "        \n",
    "            tax = f'OTU_{count}_{blasting_results_df[\"Species\"][av]}'\n",
    "        else:    \n",
    "            tax = blasting_results_df[tax_level][av]\n",
    "        count += 1\n",
    "        OTU_decoder['Seq'].append(av)\n",
    "        OTU_decoder['OTU_name'].append(tax)\n",
    "        if tax == 'nan':\n",
    "            \n",
    "            tax = 'No Fungi'\n",
    "\n",
    "        if tax not in data_tax.keys():\n",
    "    \n",
    "            data_tax[tax] = {col: 0 for col in mearged_pike_out.columns} \n",
    "        \n",
    "        for col in mearged_pike_out.columns:\n",
    "           \n",
    "            data_tax[tax][col] += mearged_pike_out[col][av]\n",
    "    \n",
    "    data_tax_df = processing_data_tax(data_tax)\n",
    "    \n",
    "    return data_tax_df, OTU_decoder\n",
    "    \n",
    "\n",
    "def filter_data(output, \n",
    "                dbpath,\n",
    "                mearged_pike_out,\n",
    "                taxonomy_level, \n",
    "                identity_filter=95, \n",
    "                cov_lim=60, \n",
    "                evalue_filter=1e-05):\n",
    "\n",
    "    # Creating output directory\n",
    "    try:\n",
    "        \n",
    "        os.mkdir(output)\n",
    "        \n",
    "    except FileExistsError:\n",
    "        \n",
    "        print('The output directory already exists!')\n",
    "        \n",
    "    consensus = create_fasta(output, mearged_pike_out, dbpath)\n",
    "    run_blast(dbpath, output)\n",
    "    data_tax = decode_tax(dbpath)\n",
    "    blasting_results_df = parse_blast(output, \n",
    "                                      dbpath, \n",
    "                                      data_tax, \n",
    "                                      consensus, \n",
    "                                      identity_filter, \n",
    "                                      cov_lim, \n",
    "                                      evalue_filter)\n",
    "\n",
    "  #  mearged_pike_out = filter_av(mearged_pike_out, prevalence, detection, slice)\n",
    "    data_tax_df, OTU_decoder = get_taxonomy(data_tax, \n",
    "                                            blasting_results_df, \n",
    "                                            mearged_pike_out,\n",
    "                                            taxonomy_level)\n",
    "\n",
    "    data_tax_df = data_tax_df[mearged_pike_out.columns]\n",
    "    for col in data_tax_df.columns:\n",
    "        \n",
    "        data_tax_df[col] = data_tax_df[col] / np.sum(data_tax_df[col].values)\n",
    "    \n",
    "    data_tax_df = data_tax_df.fillna(0)[mearged_pike_out.columns]   \n",
    "    data_tax_df = data_tax_df.assign(m=data_tax_df.mean(axis=1)).sort_values('m').drop('m', axis=1)\n",
    "\n",
    "\n",
    "    return data_tax_df, data_tax, blasting_results_df, DataFrame(OTU_decoder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "63629b7b-c05c-4055-8d3e-5b76804c6dcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "def get_color(obj_dict):\n",
    "    \n",
    "    color = ''\n",
    "    \n",
    "    while color not in obj_dict.values() and color == '':\n",
    "        \n",
    "        color = \"#\"+''.join([random.choice('0123456789ABCDEF') for j in range(6)])\n",
    "    \n",
    "    return color"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8c03e83-792f-465d-a12c-a0553e8b9d2f",
   "metadata": {},
   "source": [
    "MERGING ALL TABLES INTO ONE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e683f226-6318-4520-8e2e-26bf9657b6b2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mearged_otu_table = []\n",
    "\n",
    "#for kmer_type in ['_1_kmer_size', '_2_kmer_size', '_3_kmer_size', '_4_kmer_size','_5_kmer_size', '_6_kmer_size','_7_kmer_size']:\n",
    "for kmer_type in ['_k_3', '_k_4','_k_5', '_k_6','_k_7']:    \n",
    "    for sample in listdir(f'/mnt/AsusShareI2/RUNS/runs-sonec/pool_diff_kmers/16S/pool_V3_V4{kmer_type}/results/'):\n",
    "        opn_res = read_csv(f'/mnt/AsusShareI2/RUNS/runs-sonec/pool_diff_kmers/16S/pool_V3_V4{kmer_type}/results/{sample}/results.tsv', sep='\\t', index_col=0)\n",
    "        if 'Count' in opn_res.columns:\n",
    "            count = 0\n",
    "            mearged_otu_table.append(DataFrame(data=opn_res['Count'].tolist(), index=opn_res.index, columns=[sample + kmer_type]))\n",
    "            with open(f'/mnt/AsusShareI2/RUNS/runs-sonec/pool_diff_kmers/16S/pool_V3_V4{kmer_type}/{sample}.fasta', 'w') as opn_fasta:\n",
    "                for line in opn_res.index:\n",
    "                    opn_fasta.write(f'>{count}_{opn_res[\"Count\"][line]}\\n{line}\\n')\n",
    "                    count += 1\n",
    "        else:\n",
    "            print(f'Столбец \"Count\" отсутствует в результате для образца {sample + kmer_type}')\n",
    "\n",
    "mearged_otu_table = pd.concat(mearged_otu_table, axis=1).fillna(0)\n",
    "mearged_otu_table = mearged_otu_table.reindex(sorted(mearged_otu_table.columns), axis=1)\n",
    "mearged_otu_table.to_csv('pool_kmers_3_7_V3_V4_merged_otu_table_all_samples.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8cd4768-afe9-4d8b-ac0c-ce57bb5200bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mearged_otu_table.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51adec32-35ad-4197-8c15-17b970749e47",
   "metadata": {},
   "source": [
    "**WORKING WITH SILVA AND BLAST**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5a2426e7-2217-457c-8b62-1fe0dd026f63",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The output directory already exists!\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/09/2024 17:51:14\n",
      "New DB name:   /mnt/AsusShareI2/RUNS/runs-sonec/SILVA_138.1_SSURef_NR99_tax_silva.fasta\n",
      "New DB title:  /mnt/AsusShareI2/RUNS/runs-sonec/SILVA_138.1_SSURef_NR99_tax_silva.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /mnt/AsusShareI2/RUNS/runs-sonec/SILVA_138.1_SSURef_NR99_tax_silva.fasta\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 510508 sequences in 11.6813 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "510508it [00:05, 88860.91it/s]\n",
      "100%|█████████████████████████████████████████| 68/68 [00:00<00:00, 1001.76it/s]\n",
      "100%|█████████████████████████████████████████| 68/68 [00:00<00:00, 8314.75it/s]\n"
     ]
    }
   ],
   "source": [
    "output = f'/mnt/AsusShareI2/RUNS/runs-sonec/pool_diff_kmers/16S/pool_V1_V9{kmer_type}/TAXONOMY'\n",
    "dbpath = '/mnt/AsusShareI2/RUNS/runs-sonec/SILVA_138.1_SSURef_NR99_tax_silva.fasta'\n",
    "#taxonomy_level = 'Genus'\n",
    "taxonomy_level = 'Species'\n",
    "#taxonomy_level = 'OTU'\n",
    "\n",
    "data_tax_df, data_tax, blasting_results_df, OTU_decoder = filter_data(output, \n",
    "                                                                     dbpath,\n",
    "                                                                     mearged_otu_table,\n",
    "                                                                     taxonomy_level, \n",
    "                                                                     identity_filter=95, \n",
    "                                                                     cov_lim=60, \n",
    "                                                                     evalue_filter=1e-05)\n",
    "data_tax_df.to_csv('pool_kmers_3_7_V3_V4_data_tax_df_species.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "054ce333-9a74-4de7-aa48-4af083dd7a1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>V1_V9_1_3_kmer_size</th>\n",
       "      <th>V1_V9_1_4_kmer_size</th>\n",
       "      <th>V1_V9_1_5_kmer_size</th>\n",
       "      <th>V1_V9_1_6_kmer_size</th>\n",
       "      <th>V1_V9_1_7_kmer_size</th>\n",
       "      <th>V1_V9_2_3_kmer_size</th>\n",
       "      <th>V1_V9_2_4_kmer_size</th>\n",
       "      <th>V1_V9_2_5_kmer_size</th>\n",
       "      <th>V1_V9_2_6_kmer_size</th>\n",
       "      <th>V1_V9_2_7_kmer_size</th>\n",
       "      <th>...</th>\n",
       "      <th>V1_V9_R2_3_kmer_size</th>\n",
       "      <th>V1_V9_R2_4_kmer_size</th>\n",
       "      <th>V1_V9_R2_5_kmer_size</th>\n",
       "      <th>V1_V9_R2_6_kmer_size</th>\n",
       "      <th>V1_V9_R2_7_kmer_size</th>\n",
       "      <th>V1_V9_R3_3_kmer_size</th>\n",
       "      <th>V1_V9_R3_4_kmer_size</th>\n",
       "      <th>V1_V9_R3_5_kmer_size</th>\n",
       "      <th>V1_V9_R3_6_kmer_size</th>\n",
       "      <th>V1_V9_R3_7_kmer_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ceuthophilus sp.</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Yersinia pestis</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002201</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uncultured organism</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007767</td>\n",
       "      <td>0.003540</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uncultured bacterium</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002419</td>\n",
       "      <td>0.012719</td>\n",
       "      <td>0.010936</td>\n",
       "      <td>0.012793</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017719</td>\n",
       "      <td>0.021677</td>\n",
       "      <td>0.019287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Staphylococcus epidermidis</th>\n",
       "      <td>0.003442</td>\n",
       "      <td>0.004659</td>\n",
       "      <td>0.005174</td>\n",
       "      <td>0.004716</td>\n",
       "      <td>0.004726</td>\n",
       "      <td>0.005200</td>\n",
       "      <td>0.005923</td>\n",
       "      <td>0.006401</td>\n",
       "      <td>0.006412</td>\n",
       "      <td>0.005933</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006182</td>\n",
       "      <td>0.007066</td>\n",
       "      <td>0.006983</td>\n",
       "      <td>0.007022</td>\n",
       "      <td>0.006955</td>\n",
       "      <td>0.006835</td>\n",
       "      <td>0.007960</td>\n",
       "      <td>0.008328</td>\n",
       "      <td>0.006991</td>\n",
       "      <td>0.006857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Shigella boydii</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033975</td>\n",
       "      <td>0.041562</td>\n",
       "      <td>0.041091</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029603</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026595</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003779</td>\n",
       "      <td>0.003736</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus thermophilus</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.129145</td>\n",
       "      <td>0.128969</td>\n",
       "      <td>0.130889</td>\n",
       "      <td>0.128611</td>\n",
       "      <td>0.128807</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002734</td>\n",
       "      <td>0.002732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Yersinia pseudotuberculosis</th>\n",
       "      <td>0.083614</td>\n",
       "      <td>0.082129</td>\n",
       "      <td>0.083233</td>\n",
       "      <td>0.084282</td>\n",
       "      <td>0.084465</td>\n",
       "      <td>0.094029</td>\n",
       "      <td>0.095933</td>\n",
       "      <td>0.097887</td>\n",
       "      <td>0.096183</td>\n",
       "      <td>0.096330</td>\n",
       "      <td>...</td>\n",
       "      <td>0.075247</td>\n",
       "      <td>0.077413</td>\n",
       "      <td>0.081489</td>\n",
       "      <td>0.076058</td>\n",
       "      <td>0.079240</td>\n",
       "      <td>0.079598</td>\n",
       "      <td>0.079896</td>\n",
       "      <td>0.072343</td>\n",
       "      <td>0.074908</td>\n",
       "      <td>0.078359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Escherichia coli</th>\n",
       "      <td>0.125040</td>\n",
       "      <td>0.127229</td>\n",
       "      <td>0.091055</td>\n",
       "      <td>0.084802</td>\n",
       "      <td>0.083023</td>\n",
       "      <td>0.115013</td>\n",
       "      <td>0.115535</td>\n",
       "      <td>0.105096</td>\n",
       "      <td>0.115298</td>\n",
       "      <td>0.114434</td>\n",
       "      <td>...</td>\n",
       "      <td>0.091795</td>\n",
       "      <td>0.057041</td>\n",
       "      <td>0.082861</td>\n",
       "      <td>0.059964</td>\n",
       "      <td>0.083960</td>\n",
       "      <td>0.079635</td>\n",
       "      <td>0.083392</td>\n",
       "      <td>0.081368</td>\n",
       "      <td>0.076684</td>\n",
       "      <td>0.082172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus salivarius</th>\n",
       "      <td>0.126041</td>\n",
       "      <td>0.125020</td>\n",
       "      <td>0.124830</td>\n",
       "      <td>0.124366</td>\n",
       "      <td>0.124675</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.127451</td>\n",
       "      <td>0.124459</td>\n",
       "      <td>0.119708</td>\n",
       "      <td>0.118499</td>\n",
       "      <td>0.118425</td>\n",
       "      <td>0.120686</td>\n",
       "      <td>0.120179</td>\n",
       "      <td>0.115925</td>\n",
       "      <td>0.115710</td>\n",
       "      <td>0.115100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Klebsiella pneumoniae</th>\n",
       "      <td>0.133285</td>\n",
       "      <td>0.133373</td>\n",
       "      <td>0.132852</td>\n",
       "      <td>0.132358</td>\n",
       "      <td>0.132644</td>\n",
       "      <td>0.125780</td>\n",
       "      <td>0.124084</td>\n",
       "      <td>0.123741</td>\n",
       "      <td>0.123908</td>\n",
       "      <td>0.124098</td>\n",
       "      <td>...</td>\n",
       "      <td>0.082178</td>\n",
       "      <td>0.084161</td>\n",
       "      <td>0.089594</td>\n",
       "      <td>0.089480</td>\n",
       "      <td>0.086816</td>\n",
       "      <td>0.094278</td>\n",
       "      <td>0.096820</td>\n",
       "      <td>0.096005</td>\n",
       "      <td>0.092258</td>\n",
       "      <td>0.094713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acinetobacter radioresistens</th>\n",
       "      <td>0.131844</td>\n",
       "      <td>0.129197</td>\n",
       "      <td>0.132170</td>\n",
       "      <td>0.131639</td>\n",
       "      <td>0.131924</td>\n",
       "      <td>0.121253</td>\n",
       "      <td>0.121031</td>\n",
       "      <td>0.123182</td>\n",
       "      <td>0.121038</td>\n",
       "      <td>0.121223</td>\n",
       "      <td>...</td>\n",
       "      <td>0.105033</td>\n",
       "      <td>0.106761</td>\n",
       "      <td>0.102313</td>\n",
       "      <td>0.101348</td>\n",
       "      <td>0.101223</td>\n",
       "      <td>0.114186</td>\n",
       "      <td>0.110024</td>\n",
       "      <td>0.109615</td>\n",
       "      <td>0.109496</td>\n",
       "      <td>0.108463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactococcus lactis</th>\n",
       "      <td>0.118676</td>\n",
       "      <td>0.119558</td>\n",
       "      <td>0.119454</td>\n",
       "      <td>0.119011</td>\n",
       "      <td>0.119228</td>\n",
       "      <td>0.130246</td>\n",
       "      <td>0.129763</td>\n",
       "      <td>0.132070</td>\n",
       "      <td>0.129771</td>\n",
       "      <td>0.129969</td>\n",
       "      <td>...</td>\n",
       "      <td>0.139440</td>\n",
       "      <td>0.139292</td>\n",
       "      <td>0.134048</td>\n",
       "      <td>0.133971</td>\n",
       "      <td>0.133267</td>\n",
       "      <td>0.131331</td>\n",
       "      <td>0.128585</td>\n",
       "      <td>0.124656</td>\n",
       "      <td>0.124736</td>\n",
       "      <td>0.123863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bifidobacterium animalis</th>\n",
       "      <td>0.119316</td>\n",
       "      <td>0.119558</td>\n",
       "      <td>0.118171</td>\n",
       "      <td>0.118771</td>\n",
       "      <td>0.119388</td>\n",
       "      <td>0.118133</td>\n",
       "      <td>0.117855</td>\n",
       "      <td>0.116967</td>\n",
       "      <td>0.117863</td>\n",
       "      <td>0.118043</td>\n",
       "      <td>...</td>\n",
       "      <td>0.160485</td>\n",
       "      <td>0.154953</td>\n",
       "      <td>0.160234</td>\n",
       "      <td>0.160070</td>\n",
       "      <td>0.159970</td>\n",
       "      <td>0.159234</td>\n",
       "      <td>0.158490</td>\n",
       "      <td>0.156902</td>\n",
       "      <td>0.160323</td>\n",
       "      <td>0.158734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycoplasma gallisepticum</th>\n",
       "      <td>0.158742</td>\n",
       "      <td>0.159277</td>\n",
       "      <td>0.159085</td>\n",
       "      <td>0.158494</td>\n",
       "      <td>0.158837</td>\n",
       "      <td>0.161202</td>\n",
       "      <td>0.160906</td>\n",
       "      <td>0.163766</td>\n",
       "      <td>0.160916</td>\n",
       "      <td>0.161162</td>\n",
       "      <td>...</td>\n",
       "      <td>0.212189</td>\n",
       "      <td>0.216832</td>\n",
       "      <td>0.210051</td>\n",
       "      <td>0.205555</td>\n",
       "      <td>0.211079</td>\n",
       "      <td>0.214216</td>\n",
       "      <td>0.214655</td>\n",
       "      <td>0.211160</td>\n",
       "      <td>0.213480</td>\n",
       "      <td>0.211242</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              V1_V9_1_3_kmer_size  V1_V9_1_4_kmer_size  \\\n",
       "Ceuthophilus sp.                         0.000000             0.000000   \n",
       "Yersinia pestis                          0.000000             0.000000   \n",
       "uncultured organism                      0.000000             0.000000   \n",
       "uncultured bacterium                     0.000000             0.000000   \n",
       "Staphylococcus epidermidis               0.003442             0.004659   \n",
       "Shigella boydii                          0.000000             0.000000   \n",
       "Streptococcus thermophilus               0.000000             0.000000   \n",
       "Yersinia pseudotuberculosis              0.083614             0.082129   \n",
       "Escherichia coli                         0.125040             0.127229   \n",
       "Streptococcus salivarius                 0.126041             0.125020   \n",
       "Klebsiella pneumoniae                    0.133285             0.133373   \n",
       "Acinetobacter radioresistens             0.131844             0.129197   \n",
       "Lactococcus lactis                       0.118676             0.119558   \n",
       "Bifidobacterium animalis                 0.119316             0.119558   \n",
       "Mycoplasma gallisepticum                 0.158742             0.159277   \n",
       "\n",
       "                              V1_V9_1_5_kmer_size  V1_V9_1_6_kmer_size  \\\n",
       "Ceuthophilus sp.                         0.000000             0.000000   \n",
       "Yersinia pestis                          0.000000             0.000000   \n",
       "uncultured organism                      0.000000             0.000000   \n",
       "uncultured bacterium                     0.000000             0.000000   \n",
       "Staphylococcus epidermidis               0.005174             0.004716   \n",
       "Shigella boydii                          0.033975             0.041562   \n",
       "Streptococcus thermophilus               0.000000             0.000000   \n",
       "Yersinia pseudotuberculosis              0.083233             0.084282   \n",
       "Escherichia coli                         0.091055             0.084802   \n",
       "Streptococcus salivarius                 0.124830             0.124366   \n",
       "Klebsiella pneumoniae                    0.132852             0.132358   \n",
       "Acinetobacter radioresistens             0.132170             0.131639   \n",
       "Lactococcus lactis                       0.119454             0.119011   \n",
       "Bifidobacterium animalis                 0.118171             0.118771   \n",
       "Mycoplasma gallisepticum                 0.159085             0.158494   \n",
       "\n",
       "                              V1_V9_1_7_kmer_size  V1_V9_2_3_kmer_size  \\\n",
       "Ceuthophilus sp.                         0.000000             0.000000   \n",
       "Yersinia pestis                          0.000000             0.000000   \n",
       "uncultured organism                      0.000000             0.000000   \n",
       "uncultured bacterium                     0.000000             0.000000   \n",
       "Staphylococcus epidermidis               0.004726             0.005200   \n",
       "Shigella boydii                          0.041091             0.000000   \n",
       "Streptococcus thermophilus               0.000000             0.129145   \n",
       "Yersinia pseudotuberculosis              0.084465             0.094029   \n",
       "Escherichia coli                         0.083023             0.115013   \n",
       "Streptococcus salivarius                 0.124675             0.000000   \n",
       "Klebsiella pneumoniae                    0.132644             0.125780   \n",
       "Acinetobacter radioresistens             0.131924             0.121253   \n",
       "Lactococcus lactis                       0.119228             0.130246   \n",
       "Bifidobacterium animalis                 0.119388             0.118133   \n",
       "Mycoplasma gallisepticum                 0.158837             0.161202   \n",
       "\n",
       "                              V1_V9_2_4_kmer_size  V1_V9_2_5_kmer_size  \\\n",
       "Ceuthophilus sp.                         0.000000             0.000000   \n",
       "Yersinia pestis                          0.000000             0.000000   \n",
       "uncultured organism                      0.000000             0.000000   \n",
       "uncultured bacterium                     0.000000             0.000000   \n",
       "Staphylococcus epidermidis               0.005923             0.006401   \n",
       "Shigella boydii                          0.000000             0.000000   \n",
       "Streptococcus thermophilus               0.128969             0.130889   \n",
       "Yersinia pseudotuberculosis              0.095933             0.097887   \n",
       "Escherichia coli                         0.115535             0.105096   \n",
       "Streptococcus salivarius                 0.000000             0.000000   \n",
       "Klebsiella pneumoniae                    0.124084             0.123741   \n",
       "Acinetobacter radioresistens             0.121031             0.123182   \n",
       "Lactococcus lactis                       0.129763             0.132070   \n",
       "Bifidobacterium animalis                 0.117855             0.116967   \n",
       "Mycoplasma gallisepticum                 0.160906             0.163766   \n",
       "\n",
       "                              V1_V9_2_6_kmer_size  V1_V9_2_7_kmer_size  ...  \\\n",
       "Ceuthophilus sp.                         0.000000             0.000000  ...   \n",
       "Yersinia pestis                          0.000000             0.000000  ...   \n",
       "uncultured organism                      0.000000             0.000000  ...   \n",
       "uncultured bacterium                     0.000000             0.000000  ...   \n",
       "Staphylococcus epidermidis               0.006412             0.005933  ...   \n",
       "Shigella boydii                          0.000000             0.000000  ...   \n",
       "Streptococcus thermophilus               0.128611             0.128807  ...   \n",
       "Yersinia pseudotuberculosis              0.096183             0.096330  ...   \n",
       "Escherichia coli                         0.115298             0.114434  ...   \n",
       "Streptococcus salivarius                 0.000000             0.000000  ...   \n",
       "Klebsiella pneumoniae                    0.123908             0.124098  ...   \n",
       "Acinetobacter radioresistens             0.121038             0.121223  ...   \n",
       "Lactococcus lactis                       0.129771             0.129969  ...   \n",
       "Bifidobacterium animalis                 0.117863             0.118043  ...   \n",
       "Mycoplasma gallisepticum                 0.160916             0.161162  ...   \n",
       "\n",
       "                              V1_V9_R2_3_kmer_size  V1_V9_R2_4_kmer_size  \\\n",
       "Ceuthophilus sp.                          0.000000              0.000000   \n",
       "Yersinia pestis                           0.000000              0.000000   \n",
       "uncultured organism                       0.000000              0.000000   \n",
       "uncultured bacterium                      0.000000              0.002419   \n",
       "Staphylococcus epidermidis                0.006182              0.007066   \n",
       "Shigella boydii                           0.000000              0.029603   \n",
       "Streptococcus thermophilus                0.000000              0.000000   \n",
       "Yersinia pseudotuberculosis               0.075247              0.077413   \n",
       "Escherichia coli                          0.091795              0.057041   \n",
       "Streptococcus salivarius                  0.127451              0.124459   \n",
       "Klebsiella pneumoniae                     0.082178              0.084161   \n",
       "Acinetobacter radioresistens              0.105033              0.106761   \n",
       "Lactococcus lactis                        0.139440              0.139292   \n",
       "Bifidobacterium animalis                  0.160485              0.154953   \n",
       "Mycoplasma gallisepticum                  0.212189              0.216832   \n",
       "\n",
       "                              V1_V9_R2_5_kmer_size  V1_V9_R2_6_kmer_size  \\\n",
       "Ceuthophilus sp.                          0.000000              0.000000   \n",
       "Yersinia pestis                           0.000000              0.000000   \n",
       "uncultured organism                       0.000000              0.007767   \n",
       "uncultured bacterium                      0.012719              0.010936   \n",
       "Staphylococcus epidermidis                0.006983              0.007022   \n",
       "Shigella boydii                           0.000000              0.026595   \n",
       "Streptococcus thermophilus                0.000000              0.002734   \n",
       "Yersinia pseudotuberculosis               0.081489              0.076058   \n",
       "Escherichia coli                          0.082861              0.059964   \n",
       "Streptococcus salivarius                  0.119708              0.118499   \n",
       "Klebsiella pneumoniae                     0.089594              0.089480   \n",
       "Acinetobacter radioresistens              0.102313              0.101348   \n",
       "Lactococcus lactis                        0.134048              0.133971   \n",
       "Bifidobacterium animalis                  0.160234              0.160070   \n",
       "Mycoplasma gallisepticum                  0.210051              0.205555   \n",
       "\n",
       "                              V1_V9_R2_7_kmer_size  V1_V9_R3_3_kmer_size  \\\n",
       "Ceuthophilus sp.                          0.000000              0.000000   \n",
       "Yersinia pestis                           0.000000              0.000000   \n",
       "uncultured organism                       0.003540              0.000000   \n",
       "uncultured bacterium                      0.012793              0.000000   \n",
       "Staphylococcus epidermidis                0.006955              0.006835   \n",
       "Shigella boydii                           0.000000              0.000000   \n",
       "Streptococcus thermophilus                0.002732              0.000000   \n",
       "Yersinia pseudotuberculosis               0.079240              0.079598   \n",
       "Escherichia coli                          0.083960              0.079635   \n",
       "Streptococcus salivarius                  0.118425              0.120686   \n",
       "Klebsiella pneumoniae                     0.086816              0.094278   \n",
       "Acinetobacter radioresistens              0.101223              0.114186   \n",
       "Lactococcus lactis                        0.133267              0.131331   \n",
       "Bifidobacterium animalis                  0.159970              0.159234   \n",
       "Mycoplasma gallisepticum                  0.211079              0.214216   \n",
       "\n",
       "                              V1_V9_R3_4_kmer_size  V1_V9_R3_5_kmer_size  \\\n",
       "Ceuthophilus sp.                          0.000000              0.000000   \n",
       "Yersinia pestis                           0.000000              0.002201   \n",
       "uncultured organism                       0.000000              0.000000   \n",
       "uncultured bacterium                      0.000000              0.017719   \n",
       "Staphylococcus epidermidis                0.007960              0.008328   \n",
       "Shigella boydii                           0.000000              0.003779   \n",
       "Streptococcus thermophilus                0.000000              0.000000   \n",
       "Yersinia pseudotuberculosis               0.079896              0.072343   \n",
       "Escherichia coli                          0.083392              0.081368   \n",
       "Streptococcus salivarius                  0.120179              0.115925   \n",
       "Klebsiella pneumoniae                     0.096820              0.096005   \n",
       "Acinetobacter radioresistens              0.110024              0.109615   \n",
       "Lactococcus lactis                        0.128585              0.124656   \n",
       "Bifidobacterium animalis                  0.158490              0.156902   \n",
       "Mycoplasma gallisepticum                  0.214655              0.211160   \n",
       "\n",
       "                              V1_V9_R3_6_kmer_size  V1_V9_R3_7_kmer_size  \n",
       "Ceuthophilus sp.                          0.000000              0.001210  \n",
       "Yersinia pestis                           0.000000              0.000000  \n",
       "uncultured organism                       0.000000              0.000000  \n",
       "uncultured bacterium                      0.021677              0.019287  \n",
       "Staphylococcus epidermidis                0.006991              0.006857  \n",
       "Shigella boydii                           0.003736              0.000000  \n",
       "Streptococcus thermophilus                0.000000              0.000000  \n",
       "Yersinia pseudotuberculosis               0.074908              0.078359  \n",
       "Escherichia coli                          0.076684              0.082172  \n",
       "Streptococcus salivarius                  0.115710              0.115100  \n",
       "Klebsiella pneumoniae                     0.092258              0.094713  \n",
       "Acinetobacter radioresistens              0.109496              0.108463  \n",
       "Lactococcus lactis                        0.124736              0.123863  \n",
       "Bifidobacterium animalis                  0.160323              0.158734  \n",
       "Mycoplasma gallisepticum                  0.213480              0.211242  \n",
       "\n",
       "[15 rows x 30 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_tax_df.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "d91184c9-295a-49c8-b420-a0cd9c05bfd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>V1_V9_1_3_kmer_size</th>\n",
       "      <th>V1_V9_1_4_kmer_size</th>\n",
       "      <th>V1_V9_1_5_kmer_size</th>\n",
       "      <th>V1_V9_1_6_kmer_size</th>\n",
       "      <th>V1_V9_1_7_kmer_size</th>\n",
       "      <th>V1_V9_2_3_kmer_size</th>\n",
       "      <th>V1_V9_2_4_kmer_size</th>\n",
       "      <th>V1_V9_2_5_kmer_size</th>\n",
       "      <th>V1_V9_2_6_kmer_size</th>\n",
       "      <th>V1_V9_2_7_kmer_size</th>\n",
       "      <th>...</th>\n",
       "      <th>V1_V9_R2_3_kmer_size</th>\n",
       "      <th>V1_V9_R2_4_kmer_size</th>\n",
       "      <th>V1_V9_R2_5_kmer_size</th>\n",
       "      <th>V1_V9_R2_6_kmer_size</th>\n",
       "      <th>V1_V9_R2_7_kmer_size</th>\n",
       "      <th>V1_V9_R3_3_kmer_size</th>\n",
       "      <th>V1_V9_R3_4_kmer_size</th>\n",
       "      <th>V1_V9_R3_5_kmer_size</th>\n",
       "      <th>V1_V9_R3_6_kmer_size</th>\n",
       "      <th>V1_V9_R3_7_kmer_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>18</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   V1_V9_1_3_kmer_size  V1_V9_1_4_kmer_size  V1_V9_1_5_kmer_size  \\\n",
       "0                    9                   15                   14   \n",
       "\n",
       "   V1_V9_1_6_kmer_size  V1_V9_1_7_kmer_size  V1_V9_2_3_kmer_size  \\\n",
       "0                   11                   11                    9   \n",
       "\n",
       "   V1_V9_2_4_kmer_size  V1_V9_2_5_kmer_size  V1_V9_2_6_kmer_size  \\\n",
       "0                   12                   13                   10   \n",
       "\n",
       "   V1_V9_2_7_kmer_size  ...  V1_V9_R2_3_kmer_size  V1_V9_R2_4_kmer_size  \\\n",
       "0                   11  ...                     9                    17   \n",
       "\n",
       "   V1_V9_R2_5_kmer_size  V1_V9_R2_6_kmer_size  V1_V9_R2_7_kmer_size  \\\n",
       "0                    16                    18                    16   \n",
       "\n",
       "   V1_V9_R3_3_kmer_size  V1_V9_R3_4_kmer_size  V1_V9_R3_5_kmer_size  \\\n",
       "0                     9                    12                    19   \n",
       "\n",
       "   V1_V9_R3_6_kmer_size  V1_V9_R3_7_kmer_size  \n",
       "0                    19                    18  \n",
       "\n",
       "[1 rows x 30 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_tax_no0_rows = (data_tax_df !=0).sum()\n",
    "result = pd.DataFrame(data_tax_no0_rows).transpose()\n",
    "result.head(10)\n",
    "#result.index = ['OTU_count_nonzero']\n",
    "#result.to_csv('OTU_count_nonzero.csv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "edeae6ad-a480-4e58-8bcd-4a3cdd370229",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>column</th>\n",
       "      <th>OTU_count_nonzero</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>V1_V9_1_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>V1_V9_1_4_kmer_size</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>V1_V9_1_5_kmer_size</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>V1_V9_1_6_kmer_size</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>V1_V9_1_7_kmer_size</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>V1_V9_2_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>V1_V9_2_4_kmer_size</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>V1_V9_2_5_kmer_size</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>V1_V9_2_6_kmer_size</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>V1_V9_2_7_kmer_size</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>V1_V9_3_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>V1_V9_3_4_kmer_size</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>V1_V9_3_5_kmer_size</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>V1_V9_3_6_kmer_size</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>V1_V9_3_7_kmer_size</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>V1_V9_R1_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>V1_V9_R1_4_kmer_size</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>V1_V9_R1_5_kmer_size</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>V1_V9_R1_6_kmer_size</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>V1_V9_R1_7_kmer_size</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>V1_V9_R2_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>V1_V9_R2_4_kmer_size</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>V1_V9_R2_5_kmer_size</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>V1_V9_R2_6_kmer_size</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>V1_V9_R2_7_kmer_size</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>V1_V9_R3_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>V1_V9_R3_4_kmer_size</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>V1_V9_R3_5_kmer_size</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>V1_V9_R3_6_kmer_size</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>V1_V9_R3_7_kmer_size</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  column  OTU_count_nonzero\n",
       "0    V1_V9_1_3_kmer_size                  9\n",
       "1    V1_V9_1_4_kmer_size                 15\n",
       "2    V1_V9_1_5_kmer_size                 14\n",
       "3    V1_V9_1_6_kmer_size                 11\n",
       "4    V1_V9_1_7_kmer_size                 11\n",
       "5    V1_V9_2_3_kmer_size                  9\n",
       "6    V1_V9_2_4_kmer_size                 12\n",
       "7    V1_V9_2_5_kmer_size                 13\n",
       "8    V1_V9_2_6_kmer_size                 10\n",
       "9    V1_V9_2_7_kmer_size                 11\n",
       "10   V1_V9_3_3_kmer_size                  9\n",
       "11   V1_V9_3_4_kmer_size                 16\n",
       "12   V1_V9_3_5_kmer_size                 13\n",
       "13   V1_V9_3_6_kmer_size                 10\n",
       "14   V1_V9_3_7_kmer_size                  9\n",
       "15  V1_V9_R1_3_kmer_size                  9\n",
       "16  V1_V9_R1_4_kmer_size                 13\n",
       "17  V1_V9_R1_5_kmer_size                 17\n",
       "18  V1_V9_R1_6_kmer_size                 18\n",
       "19  V1_V9_R1_7_kmer_size                 18\n",
       "20  V1_V9_R2_3_kmer_size                  9\n",
       "21  V1_V9_R2_4_kmer_size                 17\n",
       "22  V1_V9_R2_5_kmer_size                 16\n",
       "23  V1_V9_R2_6_kmer_size                 18\n",
       "24  V1_V9_R2_7_kmer_size                 16\n",
       "25  V1_V9_R3_3_kmer_size                  9\n",
       "26  V1_V9_R3_4_kmer_size                 12\n",
       "27  V1_V9_R3_5_kmer_size                 19\n",
       "28  V1_V9_R3_6_kmer_size                 19\n",
       "29  V1_V9_R3_7_kmer_size                 18"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_long = result.melt(var_name='column', value_name='OTU_count_nonzero')\n",
    "result_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "ec94270d-2d88-48fa-aac9-b7024b031d3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>column</th>\n",
       "      <th>OTU_count_nonzero</th>\n",
       "      <th>sample_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>V1_V9_1_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>V1_V9_1_4_kmer_size</td>\n",
       "      <td>15</td>\n",
       "      <td>V1_V9_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>V1_V9_1_5_kmer_size</td>\n",
       "      <td>14</td>\n",
       "      <td>V1_V9_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>V1_V9_1_6_kmer_size</td>\n",
       "      <td>11</td>\n",
       "      <td>V1_V9_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>V1_V9_1_7_kmer_size</td>\n",
       "      <td>11</td>\n",
       "      <td>V1_V9_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>V1_V9_2_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>V1_V9_2_4_kmer_size</td>\n",
       "      <td>12</td>\n",
       "      <td>V1_V9_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>V1_V9_2_5_kmer_size</td>\n",
       "      <td>13</td>\n",
       "      <td>V1_V9_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>V1_V9_2_6_kmer_size</td>\n",
       "      <td>10</td>\n",
       "      <td>V1_V9_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>V1_V9_2_7_kmer_size</td>\n",
       "      <td>11</td>\n",
       "      <td>V1_V9_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>V1_V9_3_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>V1_V9_3_4_kmer_size</td>\n",
       "      <td>16</td>\n",
       "      <td>V1_V9_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>V1_V9_3_5_kmer_size</td>\n",
       "      <td>13</td>\n",
       "      <td>V1_V9_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>V1_V9_3_6_kmer_size</td>\n",
       "      <td>10</td>\n",
       "      <td>V1_V9_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>V1_V9_3_7_kmer_size</td>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>V1_V9_R1_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>V1_V9_R1_4_kmer_size</td>\n",
       "      <td>13</td>\n",
       "      <td>V1_V9_R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>V1_V9_R1_5_kmer_size</td>\n",
       "      <td>17</td>\n",
       "      <td>V1_V9_R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>V1_V9_R1_6_kmer_size</td>\n",
       "      <td>18</td>\n",
       "      <td>V1_V9_R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>V1_V9_R1_7_kmer_size</td>\n",
       "      <td>18</td>\n",
       "      <td>V1_V9_R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>V1_V9_R2_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>V1_V9_R2_4_kmer_size</td>\n",
       "      <td>17</td>\n",
       "      <td>V1_V9_R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>V1_V9_R2_5_kmer_size</td>\n",
       "      <td>16</td>\n",
       "      <td>V1_V9_R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>V1_V9_R2_6_kmer_size</td>\n",
       "      <td>18</td>\n",
       "      <td>V1_V9_R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>V1_V9_R2_7_kmer_size</td>\n",
       "      <td>16</td>\n",
       "      <td>V1_V9_R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>V1_V9_R3_3_kmer_size</td>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_R3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>V1_V9_R3_4_kmer_size</td>\n",
       "      <td>12</td>\n",
       "      <td>V1_V9_R3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>V1_V9_R3_5_kmer_size</td>\n",
       "      <td>19</td>\n",
       "      <td>V1_V9_R3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>V1_V9_R3_6_kmer_size</td>\n",
       "      <td>19</td>\n",
       "      <td>V1_V9_R3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>V1_V9_R3_7_kmer_size</td>\n",
       "      <td>18</td>\n",
       "      <td>V1_V9_R3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  column  OTU_count_nonzero sample_name\n",
       "0    V1_V9_1_3_kmer_size                  9     V1_V9_1\n",
       "1    V1_V9_1_4_kmer_size                 15     V1_V9_1\n",
       "2    V1_V9_1_5_kmer_size                 14     V1_V9_1\n",
       "3    V1_V9_1_6_kmer_size                 11     V1_V9_1\n",
       "4    V1_V9_1_7_kmer_size                 11     V1_V9_1\n",
       "5    V1_V9_2_3_kmer_size                  9     V1_V9_2\n",
       "6    V1_V9_2_4_kmer_size                 12     V1_V9_2\n",
       "7    V1_V9_2_5_kmer_size                 13     V1_V9_2\n",
       "8    V1_V9_2_6_kmer_size                 10     V1_V9_2\n",
       "9    V1_V9_2_7_kmer_size                 11     V1_V9_2\n",
       "10   V1_V9_3_3_kmer_size                  9     V1_V9_3\n",
       "11   V1_V9_3_4_kmer_size                 16     V1_V9_3\n",
       "12   V1_V9_3_5_kmer_size                 13     V1_V9_3\n",
       "13   V1_V9_3_6_kmer_size                 10     V1_V9_3\n",
       "14   V1_V9_3_7_kmer_size                  9     V1_V9_3\n",
       "15  V1_V9_R1_3_kmer_size                  9    V1_V9_R1\n",
       "16  V1_V9_R1_4_kmer_size                 13    V1_V9_R1\n",
       "17  V1_V9_R1_5_kmer_size                 17    V1_V9_R1\n",
       "18  V1_V9_R1_6_kmer_size                 18    V1_V9_R1\n",
       "19  V1_V9_R1_7_kmer_size                 18    V1_V9_R1\n",
       "20  V1_V9_R2_3_kmer_size                  9    V1_V9_R2\n",
       "21  V1_V9_R2_4_kmer_size                 17    V1_V9_R2\n",
       "22  V1_V9_R2_5_kmer_size                 16    V1_V9_R2\n",
       "23  V1_V9_R2_6_kmer_size                 18    V1_V9_R2\n",
       "24  V1_V9_R2_7_kmer_size                 16    V1_V9_R2\n",
       "25  V1_V9_R3_3_kmer_size                  9    V1_V9_R3\n",
       "26  V1_V9_R3_4_kmer_size                 12    V1_V9_R3\n",
       "27  V1_V9_R3_5_kmer_size                 19    V1_V9_R3\n",
       "28  V1_V9_R3_6_kmer_size                 19    V1_V9_R3\n",
       "29  V1_V9_R3_7_kmer_size                 18    V1_V9_R3"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_long['sample_name'] = result_long['column'].str.extract(r'(V1_V9_(?:R)?\\d+)_\\d+_kmer_size')\n",
    "result_long\n",
    "#result_long.to_csv('result_long_test.csv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "e359554d-bf53-4bfd-a5fb-835320a001f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OTU_count_nonzero</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>kmer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15</td>\n",
       "      <td>V1_V9_1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>V1_V9_1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>V1_V9_1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11</td>\n",
       "      <td>V1_V9_1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>12</td>\n",
       "      <td>V1_V9_2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>13</td>\n",
       "      <td>V1_V9_2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>10</td>\n",
       "      <td>V1_V9_2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>11</td>\n",
       "      <td>V1_V9_2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>16</td>\n",
       "      <td>V1_V9_3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>V1_V9_3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>10</td>\n",
       "      <td>V1_V9_3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_R1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>13</td>\n",
       "      <td>V1_V9_R1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>V1_V9_R1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>V1_V9_R1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>18</td>\n",
       "      <td>V1_V9_R1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_R2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>17</td>\n",
       "      <td>V1_V9_R2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>16</td>\n",
       "      <td>V1_V9_R2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>18</td>\n",
       "      <td>V1_V9_R2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>16</td>\n",
       "      <td>V1_V9_R2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>9</td>\n",
       "      <td>V1_V9_R3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>12</td>\n",
       "      <td>V1_V9_R3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>19</td>\n",
       "      <td>V1_V9_R3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>19</td>\n",
       "      <td>V1_V9_R3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>18</td>\n",
       "      <td>V1_V9_R3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    OTU_count_nonzero sample_name  kmer\n",
       "0                   9     V1_V9_1     3\n",
       "1                  15     V1_V9_1     4\n",
       "2                  14     V1_V9_1     5\n",
       "3                  11     V1_V9_1     6\n",
       "4                  11     V1_V9_1     7\n",
       "5                   9     V1_V9_2     3\n",
       "6                  12     V1_V9_2     4\n",
       "7                  13     V1_V9_2     5\n",
       "8                  10     V1_V9_2     6\n",
       "9                  11     V1_V9_2     7\n",
       "10                  9     V1_V9_3     3\n",
       "11                 16     V1_V9_3     4\n",
       "12                 13     V1_V9_3     5\n",
       "13                 10     V1_V9_3     6\n",
       "14                  9     V1_V9_3     7\n",
       "15                  9    V1_V9_R1     3\n",
       "16                 13    V1_V9_R1     4\n",
       "17                 17    V1_V9_R1     5\n",
       "18                 18    V1_V9_R1     6\n",
       "19                 18    V1_V9_R1     7\n",
       "20                  9    V1_V9_R2     3\n",
       "21                 17    V1_V9_R2     4\n",
       "22                 16    V1_V9_R2     5\n",
       "23                 18    V1_V9_R2     6\n",
       "24                 16    V1_V9_R2     7\n",
       "25                  9    V1_V9_R3     3\n",
       "26                 12    V1_V9_R3     4\n",
       "27                 19    V1_V9_R3     5\n",
       "28                 19    V1_V9_R3     6\n",
       "29                 18    V1_V9_R3     7"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_long['kmer'] = result_long['column'].str.extract(r'V1_V9_(?:R)?\\d+_(\\d+)_kmer_size')\n",
    "result_long.drop('column', axis=1, inplace=True)\n",
    "result_long.fillna(0)\n",
    "result_long['kmer'] = result_long['kmer'].astype('int')\n",
    "result_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "189706f3-ff41-4076-b7ed-017b494242ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_long=result_long.sort_values('kmer',ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "b75e3625-1d24-47f7-a026-df29090fed6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_long\n",
    "\n",
    "result_long.to_csv('kmers_3_7_V1_V9_OTU_count_long_table.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfe15d1d-56c8-4a35-80ab-63855eca99b5",
   "metadata": {},
   "source": [
    "**PLOTTING FROM RESULT_LONG DF( OPTIMAL)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "800c801b-27f7-4136-81b5-21b603405d17",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "for sample in result_long['sample_name'].unique():\n",
    "    subset = result_long[result_long['sample_name'] == sample]\n",
    "    ax.plot(subset['number_of_reads'], subset['OTU_count_nonzero'], marker='o', linestyle='-', label=sample)\n",
    "\n",
    "ax.set_xlabel('Number of reads')\n",
    "ax.set_ylabel('Count')\n",
    "ax.set_title('OTU')\n",
    "ax.set_ylim(-1, 35)\n",
    "ax.set_xscale('log')\n",
    "#ax.xaxis.set_ticks(result_melted[\"number_of_reads\"])\n",
    "ax.legend()\n",
    "plt.show()\n",
    "#plt.savefig('OTU_V3_V4.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc4ab5fd-eb09-4d04-a32f-f1b1da74c5c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = pd.Series([col.rsplit('_', 2)[0] for col in result.columns]).drop_duplicates().tolist()\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d597a47e-ad59-4cda-9e76-5f931456dc77",
   "metadata": {},
   "outputs": [],
   "source": [
    "reads = pd.Series([col.split('_')[-2] for col in result.columns]).drop_duplicates().tolist()\n",
    "reads.sort()\n",
    "reads"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46037c06-0247-480b-b803-79ffe57c7061",
   "metadata": {},
   "source": [
    "TESTING RESHAPING DF ON DUMMY DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "051c9bdf-a786-4ad9-8636-fc5bb17d82ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assuming your initial dataframe is named df\n",
    "data = {\n",
    "    'V3_V4_1_100_reads': [10],\n",
    "    'V3_V4_2_100_reads': [15],\n",
    "    'V3_V4_3_100_reads': [20],\n",
    "    'V3_V4_1_500_reads': [30],\n",
    "    'V3_V4_2_500_reads': [30],\n",
    "    'V3_V4_3_500_reads': [35]\n",
    "}\n",
    "\n",
    "df = pd.DataFrame(data)\n",
    "\n",
    "# Melt the dataframe to go from wide to long format\n",
    "df_long = df.melt(var_name='column', value_name='OTU_count_nonzero')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0596b50d-22bf-419b-bb6b-88cc22215e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2cc5854-be7a-41c2-aa94-3ed73594c7c4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8168a4a1-c375-4fb8-bddb-5857de263928",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract sample names and number of reads\n",
    "df_long['sample_name'] = df_long['column'].str.extract(r'(V3_V4_\\d+)_\\d+_reads')\n",
    "df_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34743d85-86a6-4517-ba7a-a9f3994af4f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_long['number_of_reads'] = df_long['column'].str.extract(r'V3_V4_\\d+_(\\d+)_reads')\n",
    "df_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c81b98e4-b3b8-4b64-8992-68f7ab1d55ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop the original column names as they are no longer needed\n",
    "df_long.drop('column', axis=1, inplace=True)\n",
    "df_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2e37c23-e272-4d5a-97e3-10c9a7f1ae85",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pivot the table to get the desired format\n",
    "df_pivot = df_long.pivot(index='number_of_reads', columns='sample_name', values='OTU_count_nonzero')\n",
    "df_pivot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aedfa68-531d-4668-96e5-001d5bdbee6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reset the index to get number_of_reads as a column\n",
    "df_final = df_pivot.reset_index()\n",
    "df_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "870cfa39-5071-4e65-9a96-528819b6ec26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Optionally, sort by number_of_reads if needed\n",
    "df_final.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01bc159a-cf66-43ec-9f48-155f470541a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_final.columns.name = None\n",
    "# Print the final dataframe\n",
    "print(df_final)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74e5dc73-f833-46f6-9f5d-8e59bd5ac64d",
   "metadata": {},
   "source": [
    "PLOTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c61ee27-6888-4b94-a2c2-c5f2f1da4efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "Color_collection = {}\n",
    "\n",
    "for i in data_tax_df.index:\n",
    "    \n",
    "    Color_collection[i] = get_color(Color_collection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c4f4352-6fa9-4483-8c7d-0c7f30dc6f94",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "import plotly.subplots as sp\n",
    "\n",
    "fig = px.bar(data_tax_df.T, \n",
    "             x=data_tax_df.columns, \n",
    "             y=data_tax_df.index,\n",
    "             width=1500, \n",
    "             height=900, \n",
    "            # color=data_tax_df.index,\n",
    "             labels={'value': 'Relative abundance', 'index':'Samples'}, \n",
    "             template='simple_white',\n",
    "             color_discrete_map=Color_collection)\n",
    "fig.update_layout(yaxis_range=[0, 1], legend_title_text='Taxon', legend_title_side='top center')\n",
    "fig.update_traces(marker_line_width=1.1, marker_line_color='#202020', opacity=0.8)\n",
    "fig.update_yaxes(ticksuffix = \"  \")\n",
    "fig.update_xaxes(range=[-1, len(mearged_otu_table.T)+0.2], autorangeoptions_clipmax=len(data_tax_df.T))\n",
    "\n",
    "#fig.update_layout(showlegend=False)\n",
    "#os.mkdir(\"VIZ\")\n",
    "fig.write_image(f\"VIZ/16S_{taxonomy_level}.pdf\")\n",
    "fig.write_image(f\"VIZ/16S_{taxonomy_level}.png\", scale=5)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15ba2a31-bac8-4851-aa4e-0ae3acd3d137",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22f05481-6c7f-40be-9c7b-56c62972ac70",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33ea1f28-e4a2-491d-90d3-9451823c5842",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
